R/secchi_workflow.R

```diff
@@ -0,0 +1,189 @@
+  ends   = format(ends, '%Y-%m-%d')
+  
+  out = list()
+  
+  for(i in seq_along(starts)){
+    cat('Downloading ', starts[i], ' to ', ends[i], '\n')
+    params$startDateLo = starts[i]
+    params$startDateHi = ends[i]
+    
+    if(starts[[i]] != ends[[i]]){
+      chunk.call = do.call(retryWQP, params)
+    
+      if(!is.null(chunk.call)){
+        meta <- attr(chunk.call, "siteInfo") %>%
+          select(MonitoringLocationIdentifier, dec_lat_va, dec_lon_va, StateCode)
+        
+        chunk.call <- rename(chunk.call, 
+                             Date=ActivityStartDate, 
+                             value=ResultMeasureValue, 
+                             units=ResultMeasure.MeasureUnitCode) %>% 
+          left_join(meta, by="MonitoringLocationIdentifier") %>%
+          rename(wqx.id=MonitoringLocationIdentifier) %>%
+          select(Date, value, units, wqx.id, dec_lat_va, dec_lon_va, StateCode)
+        
+        out[[i]] <- chunk.call       
+      }
+    }
+  }
+  
+  if(length(out) > 0){
+    out.data <- bind_rows(out)
+    
+  } else {
+    out.data <- data.frame(Date = numeric(),
+                           value=numeric(),
+                           units = character(),
+                           wqx.id = character(),
+                           dec_lat_va = numeric(),
+                           dec_lon_va = numeric(),
+                           StateCode = character())
+  }
+  
+  return(out.data)
+
+}
+
+retryWQP <- function(..., retries=3){
+  
+  safeWQP = function(...){
+    result = tryCatch({
+      readWQPdata(...)
+    }, error = function(e) {
+      if(e$message == 'Operation was aborted by an application callback'){
+        stop(e)
+      }
+      return(NULL)
+    })
+    return(result)
+  }
+  retry = 1
+  while (retry < retries){
+    result = safeWQP(...)
+    if (!is.null(result)){
+      retry = retries
+    } else {
+      message('query failed, retrying')
+      retry = retry+1
+    }
+  }
+  return(result)
+}
+
+library(dplyr)
+library(dataRetrieval)
+library(lubridate)
+library(ggplot2)
+
+characteristicNames = c("Depth, Secchi disk depth", "Depth, Secchi disk depth (choice list)", "Secchi Reading Condition (choice list)", "Secchi depth", "Water transparency, Secchi disc")
+siteTypes = "Lake, Reservoir, Impoundment"
+
+lower.48 <- stateCd$STATE[1:50]
+lower.48 <- lower.48[!(stateCd$STUSAB[1:50] %in% c("AL","HI"))]
+# To get the full data set used to produce the figure in the text:
+# get_us_secchi(outfile="all_secchi_usa.rds",
+#                   characteristicNames,
+#                   siteTypes,
+#                   stateCode = lower.48,
+#                   '1980-01-01', '2016-01-01')
+
+# To get the small subset to test the workflow:
+get_us_secchi(outfile="sub_secchi_WI_MN.rds",
+                characteristicNames,
+                siteTypes,
+                stateCode = c("01","12","27","55"), #AL, FL, WI, and MN
+                '2000-01-01', '2016-01-01')
```

Comment:
this download takes awhile, I think you should put an estimate about how long this will take as a comment